### PR TITLE
[IPsec] refactor xfrm policy state handling oss

### PR DESCRIPTION
**Pull request by ldelossa**
```release-note
Refactor XFRM policy and state creation
```

### CODE REVIEW

File: pkg/datapath/linux/ipsec.go
Comment (jschwinger233) [inline pkg/datapath/linux/ipsec.go]: nit: how about moving `if newNode.IsLocal()` from inside `enableIPSecIPvXDoLocalHost()` to here:
```
if newNode.IsLocal() {
    n.enableIPSecIPvXDoLocalHost()
}
```

Because I feel it's more natural and consistent? The subnetEncryption branch is handled same way:

```
if n.subnetEncryption() {
    return n.enableIPSecIPv4DoSubnetEncryption(newNode, nodeID, zeroMark, updateExisting, errs)
}
```

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (jschwinger233) [inline pkg/datapath/linux/ipsec/ipsec_linux.go:586]: Since dir==FWD is gone, the function name can be simplified to `_ipSecReplacePolicyIn` without a dir parameter?

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (jschwinger233) [inline pkg/datapath/linux/ipsec/ipsec_linux.go]: nit: this function looks suitable to become a method of *IPSecParameters

**Review by jschwinger233 — COMMENTED** [review]
High-level intent is clear, I left some nonblocking comments.

The commits in my opinion may not be split in a perfect way: the first commit introduces 3 functions representing 3 branches in enableIPsecIPvX(), but reviewer could have difficulty in confirming whether any logic has changed. Maybe a better way to organize these changes can be: introduce new functions commit by commit, remove the old branch replaced by new function in the same commit of new function.

New struct IPSecParameters looks good to me.

**Comment by ldelossa** [comment]
Hmm: 
> Maybe a better way to organize these changes can be: introduce new functions commit by commit, remove the old branch replaced by new function in the same commit of new function.

I don't immediately see a major difference/advantage here. The diff for the first commit only shows all green and I had thought, along with the commit message details, this made it clear that its add only. Was it not?

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (ldelossa) [reply pkg/datapath/linux/ipsec/ipsec_linux.go:586]: Yup, this is handled in a later commit, did you see this? https://github.com/cilium/cilium/pull/35210/commits/c41e71f14ee437b559e9026a8bda0823432f6f2a

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (ldelossa) [reply pkg/datapath/linux/ipsec/ipsec_linux.go]: I don't disagree, but this PR already makes a lot of changes. So may hold this one off until a follow-up.

File: pkg/datapath/linux/ipsec.go
Comment (ldelossa) [reply pkg/datapath/linux/ipsec.go]: Updated.

**Review by jschwinger233 — APPROVED** [review]
Thank you!

**Comment by joestringer** [comment]
I've switched up the release note categorization as this is not intended to have user-facing impact. Thanks!

**Comment by joestringer** [comment]
FWIW I do think that if we end up hitting a regression due to the refactoring, then it's going to be a lot more difficult to trace the changes if the commits introduce the brand new version then switch and remove the old version. I understand the nuances that in some cases it's not possible to retain a clear translation from old to new. I won't block the PR on this, but I did want to affirm one of the ideas where @jschwinger233 is coming from.

**Comment by ldelossa** [comment]
I get your point, @joestringer but I also get push back on "hard to read" pull requests, asking to split things up as done in this PR. So its seems like a point of contention. This is specifically difficult with refactors, where existing lines are being changed.

File: pkg/datapath/linux/ipsec.go
Comment (pchaigno) [inline pkg/datapath/linux/ipsec.go]: This condition is already implemented by `errors.Join()`: https://pkg.go.dev/errors#Join.

File: pkg/datapath/linux/ipsec.go
Comment (pchaigno) [inline pkg/datapath/linux/ipsec.go:481]: In commit `ipsec: make fwd policy creation explicit`, I see you're adding four `IpSecReplacePolicyFwd` statements in `pkg/datapath/linux/ipsec.go`. For `enableIPSecIPv{4,6}Do` and for `enableIPSecIPv6DoSubnetEncryption`, but not for `enableIPSecIPv4DoSubnetEncryption`. Makes sense because `enableIPSecIPv4DoSubnetEncryption` already had it. Does that mean we had a potential bug on the IPv6 codepath before due to the missing fwd policy?

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (pchaigno) [inline pkg/datapath/linux/ipsec/ipsec_linux.go]: Not sure what you mean by "used for bookkeeping"?

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (pchaigno) [inline pkg/datapath/linux/ipsec/ipsec_linux.go:127]: Could we make this higher-level? I think this one could be a boolean `EncryptedOverlay` with an appropriate description.

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (pchaigno) [inline pkg/datapath/linux/ipsec/ipsec_linux.go:51]: FYI, https://github.com/cilium/cilium/pull/35346. Although that shouldn't be a concern as I'd expect your PR to be merged first.

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (pchaigno) [inline pkg/datapath/linux/ipsec/ipsec_linux.go:265]: Commit `ipsec: remove direction from getNodeIPsecKey`:

Clearly, this is a functional change. But you're saying the PR doesn't change the behavior here (or tests wouldn't pass anyway). So I assume that means a previous commit included the inverse change in the caller. AFAIU, that means the previous commit includes a silent, breaking change. Could we avoid it and ensure no commit breaks IPsec (or if they do, at least call it out explicitly)?

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (pchaigno) [inline pkg/datapath/linux/ipsec/ipsec_linux.go]: Commit `ipsec: use appropriate name for policy in creation function`:

I'm all for small commits, but I'm going to have to agree with Gray on this one: this is just bugging reviewers when we review the previous commit and notice the function wasn't renamed then. Plus, extracting those three lines doesn't exactly simplify the previous commit.

File: pkg/datapath/linux/ipsec/ipsec_linux_test.go
Comment (pchaigno) [inline pkg/datapath/linux/ipsec/ipsec_linux_test.go]: On commit `ipsec: update IPSec tests to use declarative XFRM creation`:

Since we're on the topic of the perfect commit history:
> Additionally, existing tests are updated to utilize the structure.

This first word, in a commit description, is usually a good sign that the commit could be split :stuck_out_tongue_winking_eye:

**Review by pchaigno — COMMENTED** [review]
Mostly nits, two small concerns below:
- I'm wondering if one commit isn't actually a bugfix for IPv6.
- One commit seems to break the key derivation, then another fixes it, all silently :grimacing:

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (ldelossa) [reply pkg/datapath/linux/ipsec/ipsec_linux.go:127]: IMO, these parameters should more or less plainly spell out how the resulting policy/state looks. I don't want to have hidden resulting  values in the policy which can't be seen from the high level structure. 

But even without this concern, in a future update, we will be removing the encrypted overlay distinction all together. It's kinda a hack which we have a solution for. Happy to talk more in detail about this in real time if you'd like.

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (ldelossa) [reply pkg/datapath/linux/ipsec/ipsec_linux.go]: This one was not done for the same reason as splitting the first two commits. This, i just saw later on after I finished the bulk of this change. I'll collapse this. 👍

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (ldelossa) [reply pkg/datapath/linux/ipsec/ipsec_linux.go:265]: Yup that's fair, I caught this while testing. I can collapse this into into a relevant previous commit 👍

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (pchaigno) [reply pkg/datapath/linux/ipsec/ipsec_linux.go:127]: Sure. I only commented this because you wrote:
> Having a well defined structure allows us to declare exactly how an XFRM
policy should look at a higher level

But I don't mind either way :smiley_cat:

File: pkg/datapath/linux/ipsec.go
Comment (ldelossa) [reply pkg/datapath/linux/ipsec.go]: Aha, the IPv6 branch had this, and I copied it from there thinking we may have missed there. I'll update both IPv4 and IPv6 functions to reflect this.

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (ldelossa) [reply pkg/datapath/linux/ipsec/ipsec_linux.go]: Yeah, I just swept that explanation under the rug since it doesn't deal directly with the resulting XFRM state or policy parameters. 

I'll update the wording on this a bit better. By bookkeeping I meant, more of the administrative task of whether or not to create a policy.

Comment (pchaigno) [reply pkg/datapath/linux/ipsec/ipsec_linux.go]: It's also used for the key derivation.

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (ldelossa) [reply pkg/datapath/linux/ipsec/ipsec_linux.go]: Fixed with latest push

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (ldelossa) [reply pkg/datapath/linux/ipsec/ipsec_linux.go:265]: Fixed with latest push

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (ldelossa) [reply pkg/datapath/linux/ipsec/ipsec_linux.go]: Fixed with latest push

File: pkg/datapath/linux/ipsec.go
Comment (ldelossa) [reply pkg/datapath/linux/ipsec.go]: Fixed with latest push

File: pkg/datapath/linux/ipsec.go
Comment (ldelossa) [reply pkg/datapath/linux/ipsec.go:481]: I don't think so.

Before this PR, FWD policies were created twice (per endpoint event) for IPv4 (I guess by error), and only once for IPv6 (probably the 'right' way).

In the IPv6 case, FWD rules were created, implicitly, by creating the IN policies. 

https://github.com/cilium/cilium/blob/main/pkg/datapath/linux/ipsec.go#L387 -> https://github.com/cilium/cilium/blob/main/pkg/datapath/linux/ipsec/ipsec_linux.go#L909

Now, we just make it explicit during each node event.


As discussed over slack, in a subsequent PR we'll reduce this to just a single FWD policy being created when IPsec is enabled.

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (ldelossa) [reply pkg/datapath/linux/ipsec/ipsec_linux.go:51]: Nice! Want to push this forward after this merges? It looks good to me.

File: pkg/datapath/linux/ipsec.go
Comment (ldelossa) [reply pkg/datapath/linux/ipsec.go]: Hmm, this change seemed to break test. Let me seeee

Comment (ldelossa) [reply pkg/datapath/linux/ipsec.go]: Bleh, I see, the fix for this comment winded up creating an error **unconditionally** and adding it to the errs variable. Let me adjust..

File: pkg/datapath/linux/ipsec.go
Comment (pchaigno) [reply pkg/datapath/linux/ipsec.go:481]: I see. That makes sense. Thanks!

File: pkg/datapath/linux/ipsec.go
Comment (ldelossa) [reply pkg/datapath/linux/ipsec.go]: Fixed

**Comment by ldelossa** [comment]
@pchaigno 

Think I handled all your concerns

**Review by pchaigno — APPROVED** [review]

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (pchaigno) [reply pkg/datapath/linux/ipsec/ipsec_linux.go:265]: Nop, nop :slightly_smiling_face: 

You still needed to invert the boot IDs :wink:

**Pull request merged**
